README.md
# Подключаемся к подпольному чату

Программа для чтения и отправки сообщений в чат minecraft

![](images/chat.png)

Если вы еще не зарегистрированы в чате, программа предложит зарегистрироваться

![](images/reg.png)

## Как установить

- Для работы скрипта нужен Python версии не ниже 3.10
- Установите зависимости: `pip install -r requirements.txt`
- Если уже зарегистрированы в чате, то укажите токен в переменной окружения `TOKEN`

## Как запустить

Все аргументы запуска скриптов можно заменить настройкой в переменных окружения. Приоритет при том отдаётся аргументам запуска

```bash
python start_chat.py [-h] [--host HOST] [--reading_port READING_PORT] [--sending_port SENDING_PORT] [--token TOKEN] [--history_file HISTORY_FILE]
```

| Аргумент     | Описание                              |         По умолчанию         | Переменная окружения |
|--------------|---------------------------------------|:----------------------------:|:--------------------:|
| host         | Адрес хоста                           |      minechat.dvmn.org       |         HOST         |
| reading_port | Порт для чтения чата                  |             5000             |     READING_PORT     |
| sending_port | Порт для отправки сообщений           |             5050             |     SENDING_PORT     |
| token        | Ваш токен                             | Будет предложена регистрация |        TOKEN         |
| history_file | Имя файла для сохранения истории чата |         history.txt          |     HISTORY_FILE     |


# Цели проекта

Код написан в учебных целях — это урок в курсе по Python и веб-разработке на сайте [Devman](https://dvmn.org).

history.txt
[26.06.2023 21:34] Vlad: Какая у нее фамилия?
[26.06.2023 21:34] Eva: Ты имеешь в виду, какая у него фамилия?
[26.06.2023 21:34] Vlad: Какая фамилия у Карен?
[26.06.2023 21:34] Eva: Я не знаю Карен.


gui.py
import tkinter as tk
from asyncio import sleep
from contextlib import suppress
from tkinter import ttk
from tkinter.scrolledtext import ScrolledText

import aiofiles
from anyio import create_task_group

from mine_chat import ReadConnectionStateChanged, SendingConnectionStateChanged, NicknameReceived, MineChat


class TkAppClosed(Exception):
    pass


def process_input_text(input_field, sending_queue):
    text = input_field.get()
    sending_queue.put_nowait(text)
    input_field.delete(0, tk.END)


async def update_tk(root_frame, interval=1 / 120):
    while True:
        root_frame.event_info()
        try:
            root_frame.update()
        except tk.TclError:
            raise TkAppClosed()
        await sleep(interval)


async def update_conversation_history(panel, messages_queue, history_filepath):
    with suppress(FileNotFoundError):
        async with aiofiles.open(history_filepath, mode='r', errors='ignore', encoding='utf8') as file:
            history = await file.read()
            panel.insert('insert', history.strip())
    while True:
        msg = await messages_queue.get()

        panel['state'] = 'normal'
        if panel.index('end-1c') != '1.0':
            panel.insert('end', '\n')
        panel.insert('end', msg)
        panel.yview(tk.END)
        panel['state'] = 'disabled'


async def update_status_panel(status_labels, status_updates_queue):
    nickname_label, read_label, write_label = status_labels

    read_label['text'] = f'Чтение: нет соединения'
    write_label['text'] = f'Отправка: нет соединения'
    nickname_label['text'] = f'Имя пользователя: неизвестно'

    while True:
        msg = await status_updates_queue.get()
        if isinstance(msg, ReadConnectionStateChanged):
            read_label['text'] = f'Чтение: {msg}'

        if isinstance(msg, SendingConnectionStateChanged):
            write_label['text'] = f'Отправка: {msg}'

        if isinstance(msg, NicknameReceived):
            nickname_label['text'] = f'Имя пользователя: {msg.nickname}'


async def update_label(text_label, input_field, queue):
    while True:
        token = await queue.get()
        text_label['text'] = 'Ваш новый токен.\nСохраните эту строку в файле .env и перезапустите программу'
        input_field.insert(0, f'TOKEN = "{token}"')


def create_status_panel(root_frame):
    status_frame = tk.Frame(root_frame)
    status_frame.pack(side='bottom', fill=tk.X)

    connections_frame = tk.Frame(status_frame)
    connections_frame.pack(side='left')

    nickname_label = tk.Label(connections_frame, height=1, fg='grey', font='arial 10', anchor='w')
    nickname_label.pack(side='top', fill=tk.X)

    status_read_label = tk.Label(connections_frame, height=1, fg='grey', font='arial 10', anchor='w')
    status_read_label.pack(side='top', fill=tk.X)

    status_write_label = tk.Label(connections_frame, height=1, fg='grey', font='arial 10', anchor='w')
    status_write_label.pack(side='top', fill=tk.X)

    return nickname_label, status_read_label, status_write_label


async def draw_register_window(chat: MineChat, title: str):
    register_window = tk.Tk()
    register_window.title(title)
    register_window.geometry('500x100')
    register_window.resizable = False

    text_label = ttk.Label(register_window, font='arial 10')
    text_label.pack(expand=True)
    text_label['text'] = 'Похоже вы еще не зарегистрированы в чате.\n' \
                         'Введите никнейм для регистрации нового пользователя'

    nickname_input_frame = tk.Frame(register_window)
    nickname_input_frame.pack(side='bottom', fill=tk.X, pady=10, padx=10)
    nickname_input_field = tk.Entry(nickname_input_frame)
    nickname_input_field.pack(side='left', fill=tk.X, expand=True)
    nickname_input_field.bind('<Return>', lambda event: process_input_text(nickname_input_field, chat.nickname_queue))
    send_button = tk.Button(nickname_input_frame)
    send_button['text'] = 'Отправить'
    send_button['command'] = lambda: process_input_text(nickname_input_field, chat.nickname_queue)
    send_button.pack(side='left')

    async with create_task_group() as tg:
        tg.start_soon(chat.register_new_user)
        tg.start_soon(update_tk, register_window)
        tg.start_soon(update_label, text_label, nickname_input_field, chat.show_token_queue)


async def draw_main(chat: MineChat):
    root = tk.Tk()

    root.title('Чат Майнкрафтера')

    root_frame = tk.Frame()
    root_frame.pack(fill='both', expand=True, pady=5, padx=5)

    status_labels = create_status_panel(root_frame)

    msg_input_frame = tk.Frame(root_frame)
    msg_input_frame.pack(side='bottom', fill=tk.X)

    msg_input_field = tk.Entry(msg_input_frame)
    msg_input_field.pack(side='left', fill=tk.X, expand=True)

    msg_input_field.bind('<Return>', lambda event: process_input_text(msg_input_field, chat.sending_queue))

    send_button = tk.Button(msg_input_frame)
    send_button['text'] = 'Отправить'
    send_button['command'] = lambda: process_input_text(msg_input_field, chat.sending_queue)
    send_button.pack(side='left', pady=5, padx=5)

    conversation_panel = ScrolledText(root_frame, wrap='none')
    conversation_panel.pack(side='top', fill='both', expand=True)

    try:
        async with create_task_group() as tg:
            tg.start_soon(chat.run)
            tg.start_soon(update_tk, root_frame)
            tg.start_soon(update_conversation_history, conversation_panel, chat.messages_queue, chat.history_file)
            tg.start_soon(update_status_panel, status_labels, chat.status_updates_queue)
    finally:
        root.destroy()


mine_chat.py
import asyncio
import json
from asyncio import StreamReader, StreamWriter
from contextlib import suppress
from datetime import datetime
from enum import Enum
from json import JSONDecodeError
from typing import NoReturn, Optional

import aiofiles
from anyio import create_task_group

from errors import InvalidToken, retry_on_network_error


class MineChat:

    def __init__(
            self,
            host: str,
            reading_port: int,
            sending_port: int,
            token: str,
            history_file: str,
    ):
        self._history_file = history_file
        self._host = host
        self._listening_port = reading_port
        self._sending_port = sending_port
        self._token = token

        self.messages_queue = asyncio.Queue()
        self.saving_history_queue = asyncio.Queue()
        self.sending_queue = asyncio.Queue()
        self.status_updates_queue = asyncio.Queue()
        self.nickname_queue = asyncio.Queue()
        self.show_token_queue = asyncio.Queue()
        self.watchdog_queue = asyncio.Queue()

        self._listener: Optional[StreamReader] = None
        self._sender: Optional[StreamWriter] = None
        self._reader: Optional[StreamReader] = None
        self._writer: Optional[StreamWriter] = None

    async def run(self) -> NoReturn:
        async with create_task_group() as tg:
            tg.start_soon(self.handle_connection)

    async def register_new_user(self) -> NoReturn:
        self._token = ''
        self._reader, self._sender = await asyncio.open_connection(self._host, self._sending_port)
        try:
            async with create_task_group() as tg:
                tg.start_soon(self.log_on)
        finally:
            await self.close_connections()

    @property
    def history_file(self) -> str:
        return self._history_file

    def check_auth(self, response_text: str) -> Optional[bool]:
        with suppress(JSONDecodeError):
            response_obj = json.loads(response_text)
            if response_obj is None:
                self.watchdog_queue.put_nowait(True)
                raise InvalidToken('Неверный токен')
            elif nickname := response_obj.get('nickname'):
                if not self._token:
                    token = response_obj.get('account_hash')
                    self.show_token_queue.put_nowait(token)
                self.watchdog_queue.put_nowait(True)
                self.status_updates_queue.put_nowait(
                    NicknameReceived(nickname)
                )
                return True

    async def close_connections(self) -> None:
        with suppress(AttributeError):
            self._writer.close()
            await self._writer.wait_closed()
        with suppress(AttributeError):
            self._sender.close()
            await self._sender.wait_closed()

    @retry_on_network_error
    async def handle_connection(self) -> None:
        self.status_updates_queue.put_nowait(
            ReadConnectionStateChanged.INITIATED
        )
        self.status_updates_queue.put_nowait(
            SendingConnectionStateChanged.INITIATED
        )
        self._listener, self._writer = await asyncio.open_connection(self._host, self._listening_port)
        self._reader, self._sender = await asyncio.open_connection(self._host, self._sending_port)

        try:
            if await self.log_on():
                self.status_updates_queue.put_nowait(
                    SendingConnectionStateChanged.ESTABLISHED
                )
                async with create_task_group() as tg:
                    tg.start_soon(self.listen_chat)
                    tg.start_soon(self.save_msgs)
                    tg.start_soon(self.send_msgs)
                    tg.start_soon(self.ping_pong)
                    tg.start_soon(self.watch_for_connection)
        finally:
            await self.close_connections()

    async def listen_chat(self) -> NoReturn:
        while True:
            message = await self._listener.read(512)
            if message:
                self.watchdog_queue.put_nowait(True)
                self.status_updates_queue.put_nowait(
                    ReadConnectionStateChanged.ESTABLISHED
                )
            else:
                break
            if message_text := message.decode(errors='ignore').strip():
                receiving_time = datetime.now().strftime('%d.%m.%y %H:%M:%S')
                message_line = f'[{receiving_time}] {message_text}'
                self.messages_queue.put_nowait(message_line)
                self.saving_history_queue.put_nowait(message_line)

    async def log_on(self) -> bool:
        if self._token is None:
            raise InvalidToken('Отсутствует токен')
        expecting_auth_data = False
        while True:
            response = await self._reader.readline()
            response_text = response.decode(errors='ignore')
            self.status_updates_queue.put_nowait(
                SendingConnectionStateChanged.ESTABLISHED
            )
            if expecting_auth_data:
                if self.check_auth(response_text):
                    expecting_auth_data = False
            if 'Enter your personal hash' in response_text:
                await self.send_token()
                expecting_auth_data = True
            if 'Enter preferred nickname below' in response_text:
                await self.send_nickname()
                expecting_auth_data = True
            if 'Post your message below' in response_text:
                return True

    async def ping_pong(self) -> NoReturn:
        while True:
            self.sending_queue.put_nowait('')
            await self._reader.readline()
            self.watchdog_queue.put_nowait(True)
            await asyncio.sleep(2)

    async def save_msgs(self) -> NoReturn:
        while True:
            messages_line = await self.saving_history_queue.get()
            async with aiofiles.open(self._history_file, mode='a', errors='ignore', encoding='utf8') as file:
                await file.write(messages_line + '\n')

    async def send_msgs(self) -> NoReturn:
        while True:
            message = await self.sending_queue.get()
            message = message.replace('\n', ' ')
            self._sender.write(f'{message}\n\n'.encode(errors='ignore'))
            await self._sender.drain()
            self.watchdog_queue.put_nowait(True)

    async def send_nickname(self) -> None:
        nickname = await self.nickname_queue.get()
        self._sender.write(f'{nickname}\n'.encode(errors='ignore'))
        await self._sender.drain()

    async def send_token(self) -> None:
        self._sender.write(f'{self._token}\n'.encode(errors='ignore'))
        await self._sender.drain()
        self.watchdog_queue.put_nowait(True)

    async def watch_for_connection(self) -> NoReturn:
        while True:
            try:
                async with asyncio.timeout(3):
                    await self.watchdog_queue.get()
            except TimeoutError:
                self.status_updates_queue.put_nowait(
                    SendingConnectionStateChanged.CLOSED
                )
                self.status_updates_queue.put_nowait(
                    ReadConnectionStateChanged.CLOSED
                )
                raise ConnectionError


class ReadConnectionStateChanged(Enum):
    INITIATED = 'устанавливаем соединение'
    ESTABLISHED = 'соединение установлено'
    CLOSED = 'соединение закрыто'

    def __str__(self):
        return str(self.value)


class SendingConnectionStateChanged(Enum):
    INITIATED = 'устанавливаем соединение'
    ESTABLISHED = 'соединение установлено'
    CLOSED = 'соединение закрыто'

    def __str__(self):
        return str(self.value)


class NicknameReceived:
    def __init__(self, nickname):
        self.nickname = nickname


start_chat.py
import argparse
from _tkinter import TclError

from anyio import run
from anyio._backends._asyncio import ExceptionGroup
from environs import Env

from errors import InvalidToken
from gui import draw_main, draw_register_window
from mine_chat import MineChat


def get_args():
    env = Env()
    env.read_env()

    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--host',
        default=env('HOST', 'minechat.dvmn.org'),
        help='Host address'
    )
    parser.add_argument(
        '--reading_port',
        type=int,
        default=env.int('READING_PORT', 5000),
        help='Port for reading chat'
    )
    parser.add_argument(
        '--sending_port',
        type=int,
        default=env.int('SENDING_PORT', 5050),
        help='Port for sending messages'
    )
    parser.add_argument(
        '--token',
        default=env('TOKEN', None),
        help='Host port'
    )
    parser.add_argument(
        '--history_file',
        default=env('HISTORY_FILE', 'history.txt'),
        help='File path for saving chat history'
    )
    args = parser.parse_args()
    return args


def open_register_window(chat: MineChat, title: str):
    try:
        run(draw_register_window, chat, title)
    except (ExceptionGroup, TclError, KeyboardInterrupt):
        exit()


def main():
    args = get_args()
    chat = MineChat(
        host=args.host,
        reading_port=args.reading_port,
        sending_port=args.sending_port,
        token=args.token,
        history_file=args.history_file
    )
    try:
        run(draw_main, chat)
    except InvalidToken as ex:
        open_register_window(chat, str(ex))
    except (ExceptionGroup, TclError, KeyboardInterrupt):
        exit()


if __name__ == '__main__':
    main()


chat_reader.py
import argparse
import asyncio
import logging
from datetime import datetime
from typing import NoReturn

import aiofiles as aiofiles
from environs import Env


async def listen_chat(host: str, port: int, to_file: str) -> NoReturn:
    reader, writer = await asyncio.open_connection(host, port)
    try:
        while True:
            message = await reader.read(512)
            if not message:
                raise ConnectionError
            receiving_time = datetime.now().strftime('%d.%m.%Y %H:%M')
            message_text = message.decode(errors='ignore')
            message_line = f'[{receiving_time}] {message_text}'
            print(message_line)
            async with aiofiles.open(to_file, mode='a', errors='ignore', encoding='utf8') as file:
                logging.debug(message_line)
                await file.write(message_line)
    finally:
        writer.close()
        await writer.wait_closed()


def main():
    env = Env()
    env.read_env()

    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--host',
        default=env('HOST', 'minechat.dvmn.org'),
        help='Host address'
    )
    parser.add_argument(
        '--port',
        type=int,
        default=env.int('PORT', 5000),
        help='Host port'
    )
    parser.add_argument(
        '--to_file',
        default=env('TO_FILE', 'minechat.txt'),
        help='File name to save chat history'
    )
    parser.add_argument(
        '--log_off',
        action='store_false',
        help='Disable logging'
    )
    parser.add_argument(
        '--log_filename',
        default=env('LOG_FILENAME', default='chat.log'),
        help='Set log file name'
    )
    args = parser.parse_args()

    logging.disable(
        args.log_off or env.bool('LOG_OFF', default=False)
    )
    logging.basicConfig(
        level=logging.DEBUG,
        filename=args.log_filename
    )

    while True:
        try:
            asyncio.run(listen_chat(
                host=args.host,
                port=args.port,
                to_file=args.to_file,
            ))
        except Exception:
            continue



if __name__ == '__main__':
    main()


chat_writer.py
import argparse
import asyncio
import json
import logging
from asyncio import StreamWriter
from datetime import datetime
from functools import partial
from typing import NoReturn, Callable, Optional

from environs import Env


async def submit_message(writer: StreamWriter, message: str) -> bool:
    message = message.replace('\n', ' ')
    writer.write(f'{message}\n\n'.encode(errors='ignore'))
    await writer.drain()
    sending_time = datetime.now().strftime('%d.%m.%Y %H:%M')
    logging.debug(f'Sending message [{sending_time}] {message}')
    print('Сообщение отправлено')
    return True


async def register(writer: StreamWriter, nickname: str = None) -> None:
    if nickname:
        print(f'Регистрируемся под ником: {nickname}')
    else:
        nickname = input('Введите никнейм для регистрации:')
    nickname = nickname.replace('\n', ' ')
    writer.write(f'{nickname}\n'.encode(errors='ignore'))
    await writer.drain()


async def authorise(writer: StreamWriter, chat_token: str, ) -> None:
    writer.write(f'{chat_token}\n'.encode(errors='ignore'))
    await writer.drain()


def print_report(response: str) -> None:
    try:
        response_obj = json.loads(response)
    except json.decoder.JSONDecodeError:
        return
    else:
        if response_obj is None:
            print('Неизвестный токен.')
        elif token := response_obj.get('account_hash'):
            print(f'Ваш токен: {token}\nСохраните его в файле .env')


def get_action_func(
        response: str,
        message: str,
        chat_token: str,
        nickname: str
) -> Optional[Callable]:
    if 'Enter preferred nickname below' in response:
        return partial(register, nickname=nickname)
    if 'Enter your personal hash' in response:
        return partial(authorise, chat_token=chat_token)
    if 'Post your message below' in response:
        return partial(submit_message, message=message)


async def send_message(
        host: str,
        port: int,
        chat_token: str,
        message: str,
        nickname: str
) -> NoReturn:
    reader, writer = await asyncio.open_connection(host, port)
    try:
        while True:
            response = await reader.read(512)
            response = response.decode(errors='ignore')
            action = get_action_func(
                response,
                message,
                chat_token,
                nickname
            )
            if action:
                message_sent = await action(writer=writer)
                if message_sent:
                    break
            else:
                print_report(response)
    finally:
        writer.close()
        await writer.wait_closed()


def main():
    env = Env()
    env.read_env()

    parser = argparse.ArgumentParser()
    parser.add_argument(
        'message',
        help='Message for sending'
    )
    parser.add_argument(
        '--nickname',
        default=env('NICKNAME', default=None),
        help='Nickname for registration'
    )
    parser.add_argument(
        '--host',
        default=env('HOST', 'minechat.dvmn.org'),
        help='Host address'
    )
    parser.add_argument(
        '--port',
        type=int,
        default=env.int('PORT', 5050),
        help='Host port'
    )
    parser.add_argument(
        '--log_off',
        action='store_false',
        help='Disable logging'
    )
    parser.add_argument(
        '--log_filename',
        default=env('LOG_FILENAME', default='server.log'),
        help='Set log file name'
    )
    args = parser.parse_args()

    logging.disable(
        args.log_off or env.bool('LOG_OFF', default=False)
    )

    logging.basicConfig(
        level=logging.DEBUG,
        filename=args.log_filename
    )

    asyncio.run(send_message(
        host=args.host,
        port=args.port,
        chat_token=env('CHAT_TOKEN', default=None),
        message=args.message,
        nickname=args.nickname
    ))


if __name__ == '__main__':
    main()


main.py
import argparse
import asyncio
from asyncio import StreamWriter
from datetime import datetime
from typing import NoReturn

import aiofiles
from environs import Env

import gui


async def save_messages(
        saving_history_queue: asyncio.Queue,
        filepath: str
) -> NoReturn:
    while True:
        messages_line = await saving_history_queue.get()
        async with aiofiles.open(filepath, mode='a', errors='ignore', encoding='utf8') as file:
            await file.write(messages_line + '\n')


async def read_msgs(
        messages_queue: asyncio.Queue,
        saving_history_queue: asyncio.Queue,
        host: str,
        port: int
) -> NoReturn:
    reader, writer = await asyncio.open_connection(host, port)
    while True:
        message = await reader.read(512)
        if message_text := message.decode(errors='ignore').strip():
            receiving_time = datetime.now().strftime('%d.%m.%Y %H:%M')
            message_line = f'[{receiving_time}] {message_text}'
            messages_queue.put_nowait(message_line)
            saving_history_queue.put_nowait(message_line)


async def submit_msgs(
        writer: StreamWriter,
        sending_queue: asyncio.Queue
) -> NoReturn:
    while True:
        message = await sending_queue.get()
        message = message.replace('\n', ' ')
        writer.write(f'{message}\n\n'.encode(errors='ignore'))
        await writer.drain()


async def authorise(writer: StreamWriter, token: str) -> None:
    writer.write(f'{token}\n'.encode(errors='ignore'))
    await writer.drain()


async def handle_server_responses(
        host: str,
        sending_port: str,
        token: str,
        sending_queue: asyncio.Queue
) -> NoReturn:
    reader, writer = await asyncio.open_connection(host, sending_port)
    try:
        while True:
            response = await reader.read(512)
            response_text = response.decode(errors='ignore')
            if 'Enter your personal hash' in response_text:
                await authorise(writer, token)
            elif 'Post your message below' in response_text:
                await submit_msgs(writer, sending_queue)
            else:
                print(response_text)
    finally:
        writer.close()
        await writer.wait_closed()


async def main():
    args = get_args()

    messages_queue = asyncio.Queue()
    sending_queue = asyncio.Queue()
    status_updates_queue = asyncio.Queue()
    saving_history_queue = asyncio.Queue()

    await asyncio.gather(
        save_messages(saving_history_queue, args.history_file),
        read_msgs(messages_queue, saving_history_queue, args.host, args.reading_port),
        gui.draw(messages_queue, sending_queue, status_updates_queue, args.history_file),
        handle_server_responses(args.host, args.sending_port, args.token, sending_queue)
    )


def get_args():
    env = Env()
    env.read_env()

    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--host',
        default=env('HOST', 'minechat.dvmn.org'),
        help='Host address'
    )
    parser.add_argument(
        '--reading_port',
        type=int,
        default=env.int('READING_PORT', 5000),
        help='Port for reading chat'
    )
    parser.add_argument(
        '--sending_port',
        type=int,
        default=env.int('SENDING_PORT', 5050),
        help='Port for sending messages'
    )
    parser.add_argument(
        '--token',
        default=env('TOKEN', None),
        help='Host port'
    )
    parser.add_argument(
        '--history_file',
        default=env('HISTORY_FILE', 'history.txt'),
        help='File path for saving chat history'
    )
    args = parser.parse_args()
    return args


if __name__ == '__main__':
    asyncio.run(main())
